### DISCUSSION TITLE
Numerical Methods or Calculus: integration, differentation, interpolation, etc

### DISCUSSION THREAD

**Discussion by spencerstith** [josdejong/mathjs#2676] [Ideas]
I have been browsing this repo and noticed there doesn't seem to be any functionality for numerical methods for integration, derivatives, interpolation, etc. Is that needed in this repository? I would love to start on that unless it is beyond the scope of this library.

**Comment by Madanaa** [comment]
@spencerstith  Euler method, Second-order Runge-Kutta methods

**Comment by cshaa** [comment]
Hey Spencer,
Thanks for getting in touch! I think that numerical methods are *not* out of scope – we already have a [symbolic derivative](https://mathjs.org/docs/reference/functions/derivative.html) and a feature request for symbolic integration (#1015). Numerical integration/derivative would be a great fallback for functions that we can't handle symbolically.

By interpolation I assume you mean something like [`numpy.interp`](https://numpy.org/doc/stable/reference/generated/numpy.interp.html). In this regard, I like the idea of creating a new type `SampledFunction` which would represent points like _[f(0), f(0.1), ... f(1)]_ for some function _f_. This would make a clear distinction between vectors and sampled functions (which is missing in numpy). Then, interpolation would take a `SampledFunction` and return another, more densely populated `SampledFunction`. This would work nicely with FFT (#46) too.

I can imagine that ODE iterators like Euler and Runge-Kutta, which Madan mentions, would also be useful in math.js. But maybe after we have a solver for _algebraic_ equations, so that we can make a consistent API? 🤔 Not sure about this, I'd love to hear your thoughts!

Anyway, starting a new issue about a specific feature would be a great start. Then you (or anyone else) can propose an API and later follow up with a PR :)

**Comment by cshaa** [comment]
@josdejong Any thoughts about this, Jos?

**Comment by josdejong** [comment]
Yes, I would love to have numerical implementations to do integration, differentiation, etc 👍 . Thanks for bringing this up.

Related: #227

**Comment by jjram20** [comment]
Could I take this issue? I can help to implement a Runge-Kutta or maybe some integration algorithms.

**Comment by josdejong** [comment]
Sounds good, thanks for your offer @jjram20 👍 . 

Before you work out new functions in detail, please keep us up to date on what exactly you plan to implement and how, so we can discuss where needed.

**Comment by jjram20** [comment]
I could start creating a new folder for numerical methods for calculus inside of /src/functions including a function for numerical integration using the trapezoidal rule, the function would include as parameters the functions that will be integrated as a string and it will be evaluated using the method evaluate of this library, the other parameters will be the limits of integration and the size of increments. The idea is to evaluate using for cycles.

If I am not considering anything, am not being clear, or anything else please let me know.

**Comment by josdejong** [comment]
Yes it makes sense to introduce a new category, "numerical" or "calculus" (what would be the best name?), and create a folder in `/src/function`. What functions do you exactly have in mind? numerical integration, numerical differentiation, ...?

**Comment by jjram20** [comment]
For now I would like to start including numerical differentiation (using the definition (f(x+h)-f(x))/h) and numerical integration using trapezoidal rule, and later the next step could be to add other algorithms as Runge-Kutta, but step by step, what do you think about it?

**Comment by josdejong** [comment]
That sounds like a good and realistic plan (not too much at once). Do you have any preference for the name of the category? Ie. "numerical" or "calculus" or something else?

**Comment by jjram20** [comment]
The name of the folder can be "numerical".

**Comment by josdejong** [comment]
Yes I like that best too 👍

**Reply by dvd101x** [reply]
Hi Jos, this a topic that I'm interested in and I have something I been working for ODE with different methods Euler, RK2, RK4, etc.

Is it OK if I start to share some ideas and snippets of code in this thread?

**Reply by dvd101x** [reply]
Hi Jos, I've been working on a way to include a function `solveODE` that can solve with the Euler method like this:
$$y_{n+1} = y_n + hf(t_n,y_n)$$

```python
# solveODE( f(t, y), [t_start, t_end, delta_t], [...initial_state])

solveODE( _(t,y) = y, [0,4,1], 1)
# yields {t: [0, 1, 2, 3, 4], y: [1, 2, 4, 8, 16]}

solveODE( _(t,y) = [y[1]], [0,4,1], [1])
# yields {t: [0, 1, 2, 3, 4], y: [[1], [2], [4], [8], [16]]}

solveODE( _(t,y) = y, [0,4,1], [1, 1.1])
# yields {t: [0, 1, 2, 3, 4], "y": [[1, 1.1], [2, 2.2], [4, 4.4], [8, 8.8], [16, 17.6]]}

solveODE( _(t,y) = y 1/s, [0 s,4 s,1 s], 3 m)
#  yields {t: [0 s, 1 s, 2 s, 3 s, 4 s], y: [3 m, 6 m, 12 m, 24 m, 48 m]}

solveODE( _(t,y) = y 1/s, [0 s, 4 s, 1 s], [3 m])
# yields {t: [0 s, 1 s, 2 s, 3 s, 4 s], y: [[3 m], [6 m], [12 m], [24 m], [48 m]]}
```

To test this you can just copy paste the examples in the left panel here: https://dvd101x.github.io/Engineering-Solver/

 I could add different methods overtime:
- Euler Method
- Runge Kutta 2nd order
- Runge Kutta 4th order
- Adams Bashforth 5 steps (starting with Range Kutta 4th order)
- Dormand Prince (ode45)
- Maybe LSODA

I think I can make a PR with the Euler method in a few weeks if that's ok

These are some references:
- [ode45](https://octave.sourceforge.io/octave/function/ode45.html)
- [solve_ivp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html)

PD: it can even solve [Rocket Trajectory Optimization](examples/browser/rocket_trajectory_optimization.html)

```python
# Define initial values
G = gravitationConstant # Gravitational constant
mbody = 5.9724e24 kg    # Mass of Earth
mu = G * mbody          # Standard gravitational parameter
g0 = gravity            # Standard gravity: used for calculating propellant consumption (dmdt)
r0 = 6371 km            # Mean radius of Earth
t0 = 0 s                # Simulation start
dt = 0.5 s              # Simulation timestep
t_stage1 = 149.5 s      # Simulation duration
isp_sea = 282 s         # Specific impulse (at sea level)
isp_vac = 311 s         # Specific impulse (in vacuum)
gamma0 = 89.99970 deg   # Initial pitch angle (90 deg is vertical)
v0 = 1 m/s              # Initial velocity (must be non-zero because ODE is ill-conditioned)
phi0 = 0 deg            # Initial orbital reference angle
m1 = 433100 kg          # First stage mass
m2 = 111500 kg          # Second stage mass
m3 = 1700 kg            # Third stage / fairing mass
mp = 5000 kg            # Payload mass
m0 = m1+m2+m3+mp        # Initial mass of rocket
dm = 2750 kg/s          # Mass flow rate
A = (3.66 m)^2 * pi     # Area of the rocket
dragCoef = 0.2          # Drag coefficient

# Define the equations of motion. We just thrust into current direction of motion, e.g. making a gravity turn.
gravity(r) = mu / r.^2
angVel(r, v, gamma) = v/r * cos(gamma) * rad                          # Angular velocity of rocket around moon
density(r) = 1.2250 kg/m^3 * exp(-g0 * (r - r0) / (83246.8 m^2/s^2))  # Assume constant temperature
drag(r, v) = 1/2 * density(r) .* v.^2 * A * dragCoef

# pressure ~ density for constant temperature
isp(r) = isp_vac + (isp_sea - isp_vac) * density(r)/density(r0)
thrust(r) = g0 * isp(r) * dm
 
# It is important to maintain the same argument order for each of these functions.
drdt(v, gamma) = v sin(gamma)
dvdt(r, v, m, gamma) = - gravity(r) * sin(gamma) + (thrust(r) - drag(r, v)) / m
dmdt() = - dm
dphidt(r, v, gamma) = angVel(r, v, gamma)
dgammadt(r, v, gamma) = angVel(r, v, gamma) - gravity(r) * cos(gamma) / v * rad
dydt(t, y) = [
  drdt(y[2], y[5]),
  dvdt(y[1], y[2], y[3], y[5]),
  dmdt(),
  dphidt(y[1], y[2], y[5]),
  dgammadt(y[1], y[2], y[5])
]

# Remember to maintain the same variable order in the call to ndsolve.
x = [r0, v0, m0, phi0, gamma0]
result_stage1 = solveODE(dydt, [t0, t_stage1, dt], x)

# Reset initial conditions for interstage flight
dm = 0 kg/s
t_interstage = t_stage1 + 10 s
x = flatten(result_stage1.y[end,:])
x[3] = m2+m3+mp   # New mass after stage seperation
result_interstage = solveODE(dydt, [t_stage1, t_interstage, dt], x)

# Reset initial conditions for stage 2 flight
dm = 270.8 kg/s
isp_vac = 348 s
t_stage2 = t_interstage + 350 s
x = flatten(result_interstage.y[end,:])
result_stage2 = solveODE(dydt, [t_interstage, t_stage2, dt], x)

# Reset initial conditions for unpowered flight
dm = 0 kg / s
t_unpowered1 = t_stage2 + 900 s
dt = 10 s
x = flatten(result_stage2.y[end,:])
result_unpowered1 = solveODE(dydt, [t_stage2, t_unpowered1, dt], x)

# Reset initial conditions for final orbit insertion
dm = 270.8 kg / s
t_insertion = t_unpowered1 + 39 s
dt = 0.5 s
x = flatten(result_unpowered1.y[end,:])
result_insertion = solveODE(dydt, [t_unpowered1, t_insertion, dt], x)

# Reset initial conditions for unpowered flight
dm = 0 kg / s
t_unpowered2 = t_insertion + 250 s
dt = 10 s
x = flatten(result_insertion.y[end,:])
result_unpowered2 = solveODE(dydt, [t_insertion, t_unpowered2, dt], x)
```

**Reply by dvd101x** [reply]
Included a fourth option to optionally specify a method (default: "euler"):

```python
# solveODE( f(t, y), [t_start, t_end, delta_t], [...initial_state], ? "method")

method = "rk4" # options are [euler, rk2, ralston, rk4, ab5]

solveODE( _(t,y) = y, [0,4,1], 1, method)
# yields {t: [0, 1, 2, 3, 4], y: [1, 2.7083333333333, 7.3350694444444, 19.865813078704, 53.803243754823]}

solveODE( _(t,y) = [y[1]], [0,4,1], [1], method)
# yields {t: [0, 1, 2, 3, 4], y: [[1], [2.7083333333333], [7.3350694444444], [19.865813078704], [53.803243754823]]}

solveODE( _(t,y) = y, [0,4,1], [1, 1.1], method)
# yields {t: [0, 1, 2, 3, 4], y: [[1, 1.1], [2.7083333333333, 2.9791666666667], [7.3350694444444, 8.0685763888889], [19.865813078704, 21.852394386574], [53.803243754823, 59.183568130305]]}

solveODE( _(t,y) = y 1/s, [0 s,4 s,1 s], 1 m, method)
# yields {t: [0 s, 1 s, 2 s, 3 s, 4 s], y: [1 m, 2.7083333333333 m, 7.3350694444444 m, 19.865813078704 m, 53.803243754823 m]}

solveODE( _(t,y) = y 1/s, [0 s, 4 s, 1 s], [1 m], method)
# yields {t: [0 s, 1 s, 2 s, 3 s, 4 s], y: [[1 m], [2.7083333333333 m], [7.3350694444444 m], [19.865813078704 m], [53.803243754823 m]]}
```

**Reply by dvd101x** [reply]
I've been working on a variable step method and it seems promising.

Dormand Prince RK5(4)7M
https://www.sciencedirect.com/science/article/pii/0771050X80900133?via%3Dihub

**Reply by dvd101x** [reply]
Here is an implementation of Dormand Prince RK5(4)7M

https://observablehq.com/@dvd101x/rk54

I would like to discuss some ideas regarding the use of units and nomenclature if possible.

**Reply by dvd101x** [reply]
Here is an example for the [Lorenz attractor](https://observablehq.com/@dvd101x/lorenz-attractor) using the same algorithm.

**Reply by josdejong** [reply]
@dvd101x really nice to see you have the Euler method working, even with units! Thanks for sharing. It makes sense to implement different methods over time, since each method has pro's and con's (performance, accuracy, stability). It will be good to allow the user choose a method.

It may be good to change the signature of `solveODE` to accept an object with properties to make it easier to use, like:

```js
solveODE({
  f,
  tStart,
  tEnd,
  deltaT,
  initialState,
  method
})

// or
solveODE({
  f,
  t: start:step:end,
  initialState,
  method
})

// or something in that direction
```

A PR with the Euler method would be very welcome!

What ideas do you want to discuss regarding units?

**Reply by dvd101x** [reply]
## Nomenclature

I agree regarding the object, maybe take the default stuff that is always needed as ordered arguments and the rest inside an options object.

For fixed step methods

```javascript
solveODE(
  f,
  [tStart, tEnd],
  initialState,
  {
    method,
    step
  }
)
```

For variable step methods

```javascript
solveODE(
  f,
  [tStart, tEnd],
  initialState,
  {
    method,
    initialStep,
    minStep,
    maxStep,
    tol,
    maxIter
  }
)
```

I was able to work on different methods that seem to be working, so far I think this could be covered.

| Fixed Step | Variable Step |
|---|---|
| RK1 (Euler) | RK23 |
| RK2 (Explicit Midpoint) | RK45 (Dormand Prince) |
| RK4 (Runge Kutta 4th order) | ~~BDF (future implementation)~~|
| AB5 (Adams-Bashforth 5th order) | ~~LSODA (future implementation)~~|

## Units

Regarding units, I was wondering if it might be loosing efficiency as the units are implemented in every step. I was wondering if that's ok or if everything should be converted to values, then do every iteration and back to units for the final result.

I don't know how to do that but maybe an initial PR might work with units all the way as default and for the future think about that possibility of optimization if that makes sense.

**Reply by josdejong** [reply]
Hm, yeah. Or maybe only keep the function separate, like `solveODE(f, { tStart, tEnd, initialState, ... })`. I still have a slight preference for putting all parameters in an object over "some" mix of unnamed and named parameters and having some parameters wrapped in an array. How to remember which parameter should go in or outside of the object, at what order and whether wrapped in an array? A single object unifies this.

About units: I would say go for the easiest solution at first ("just make it work"), then see if there is need to optimize, and if so try to implement a faster solution in a second iteration (unless it is really trivial).

**Reply by dvd101x** [reply]
This is a comparison of **most parameters as an object** vs **optional parameters as an object**

```py
solveODE(f, {tStart:0, tEnd:4, initialState:1, step:1, method:'RK4'})
# vs
solveODE(f, [0, 4], 1, {step:1, method:'RK4'})
```

I was thinking about the possibility of including a default method with variable step and default tolerance if no options are available thus this would also be possible
```py
solveODE(f, {tStart:0, tEnd:4, initialState:1})
# vs
solveODE(f, [0, 4], 1)
```

This is emulating some other solvers that I know, like matlab `ode45(f, xspan, y0);` or scipy `solve_ivp(f, xspan, y0)`.

You are making some valid points regarding the mix of unnamed, named, arrays, functions, etc. If it's more inline with the rest of the library for most parameters to be as an object that's ok for me, just please review, there might be some merits in this form.

**Reply by josdejong** [reply]
An example like `solveODE(f, [0, 4], 1)` is indeed nicely compact and readable. I was triggered when first looking at your examples in https://github.com/josdejong/mathjs/discussions/2676#discussioncomment-5574215, I couldn't see the trees from the forest with all the commas and brackets and was wondering what variable is what.

Having three parameters like `ode45(f, xspan, y0)` look very clear. With JavaScript and in mathjs we can put that in an object, like `ode45({ f, xspan, y0 })`, making it all named parameters, and opening up space to introduce extra optional (named) arguments without making it confusing. It think it will be best to use a single object (matching what is common in JS and in mathjs), but we can think through how to pass start, end, and optional step in a good way (array? range? individual parameters?).

```js
solveODE({
  f,
  xspan: [start, step, end],  // or [start, end, step], or {start, step, end}, or or start:step:end
  y0
})
```

On a side note: we should really implement object literal property value shorthand in the expression parser, I use that extensively in JS. I've opened a discussion here: https://github.com/josdejong/mathjs/discussions/2953

**Reply by dvd101x** [reply]
Sorry about the confusion, I've been tweaking the nomenclature over time, specially after I saw I could make a variable step method.

I can work on a concept based on other functions that can take an object as an argument like: `derivative`, `rationalize`, `resolve`, `simplifyConstant`, `simplify`, `simplifyCore` and `symbolicEqual`. I will review those an try to come up with something.

It sounds very nice to implement #2953, I recently learned about it in js.

**Reply by josdejong** [reply]
Sounds good!

**Reply by dvd101x** [reply]
Hi Jos,

Here is an initial PR for your review

#2958
